File: backend/models/Notification.js
class Notification {
  constructor({
    id,
    userId,
    type,
    title,
    message,
    data = {},
    isRead = false,
    createdAt = new Date(),
    expiresAt = null
  }) {
    this.id = id;
    this.userId = userId;
    this.type = type; // 'report_ready', 'doctor_feedback', 'system_alert', 'appointment_reminder'
    this.title = title;
    this.message = message;
    this.data = data; // Additional data specific to notification type
    this.isRead = isRead;
    this.createdAt = createdAt;
    this.expiresAt = expiresAt;
  }

  static fromJson(json) {
    return new Notification({
      id: json.id,
      userId: json.userId,
      type: json.type,
      title: json.title,
      message: json.message,
      data: json.data || {},
      isRead: json.isRead || false,
      createdAt: new Date(json.createdAt),
      expiresAt: json.expiresAt ? new Date(json.expiresAt) : null
    });
  }

  toJson() {
    return {
      id: this.id,
      userId: this.userId,
      type: this.type,
      title: this.title,
      message: this.message,
      data: this.data,
      isRead: this.isRead,
      createdAt: this.createdAt.toISOString(),
      expiresAt: this.expiresAt ? this.expiresAt.toISOString() : null
    };
  }

  // Check if notification is expired
  isExpired() {
    if (!this.expiresAt) return false;
    return new Date() > this.expiresAt;
  }

  // Mark notification as read
  markAsRead() {
    this.isRead = true;
  }
}

module.exports = Notification;

File: frontend/lib/models/notification.dart
class Notification {
  final int id;
  final int userId;
  final String type;
  final String title;
  final String message;
  final Map<String, dynamic> data;
  final bool isRead;
  final DateTime createdAt;
  final DateTime? expiresAt;

  Notification({
    required this.id,
    required this.userId,
    required this.type,
    required this.title,
    required this.message,
    required this.data,
    required this.isRead,
    required this.createdAt,
    this.expiresAt,
  });

  factory Notification.fromJson(Map<String, dynamic> json) {
    return Notification(
      id: json['id'],
      userId: json['userId'],
      type: json['type'],
      title: json['title'],
      message: json['message'],
      data: Map<String, dynamic>.from(json['data'] ?? {}),
      isRead: json['isRead'] ?? false,
      createdAt: DateTime.parse(json['createdAt']),
      expiresAt: json['expiresAt'] != null 
          ? DateTime.parse(json['expiresAt']) 
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'userId': userId,
      'type': type,
      'title': title,
      'message': message,
      'data': data,
      'isRead': isRead,
      'createdAt': createdAt.toIso8601String(),
      'expiresAt': expiresAt?.toIso8601String(),
    };
  }

  bool get isExpired {
    if (expiresAt == null) return false;
    return DateTime.now().isAfter(expiresAt!);
  }

  Notification copyWith({
    int? id,
    int? userId,
    String? type,
    String? title,
    String? message,
    Map<String, dynamic>? data,
    bool? isRead,
    DateTime? createdAt,
    DateTime? expiresAt,
  }) {
    return Notification(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      type: type ?? this.type,
      title: title ?? this.title,
      message: message ?? this.message,
      data: data ?? this.data,
      isRead: isRead ?? this.isRead,
      createdAt: createdAt ?? this.createdAt,
      expiresAt: expiresAt ?? this.expiresAt,
    );
  }

  // Helper methods for notification types
  bool get isECGReportReady => type == 'ecg_report_ready';
  bool get isDoctorFeedback => type == 'doctor_feedback';
  bool get isSystemAlert => type == 'system_alert';
  bool get isAppointmentReminder => type == 'appointment_reminder';

  @override
  String toString() {
    return 'Notification(id: $id, type: $type, title: $title, isRead: $isRead)';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is Notification && other.id == id;
  }

  @override
  int get hashCode => id.hashCode;
}